File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/c/suffix-structures.c
/*
 * C Implementation: Suffix Structures (Suffix Array & Automaton)
 * Invariant: Substring queries in O(M) time
 */
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int* data;
    int size;
    int capacity;
} suffix_structures_t;

// TODO: Complete the allocation and operations
void init_suffix_structures(suffix_structures_t* item, int capacity) {
    item->data = (int*)malloc(capacity * sizeof(int));
    item->size = 0;
    item->capacity = capacity;
}

File: src/courses/data/catalog/02-dsa/advanced-dsa/implementations/java/suffix-structures.java
/**
 * Java Implementation: Suffix Structures (Suffix Array & Automaton)
 * Invariant: Substring queries in O(M) time
 */
package seed.dsa.core;

import java.util.*;

public class SuffixStructuresSuffixArrayAutomaton<T> {
    private final List<T> items = new ArrayList<>();

    // TODO: Implement your solution here
    public void add(T value) {
        items.add(value);
    }

    public int size() {
        return items.size();
    }
}
